presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java

```diff
@@ -458,16 +504,120 @@ else if (isCovariantParameterPosition(commonSuperTypeBase.get(), i)) {
         return Optional.of(new TypeSignature(commonSuperTypeBase.get(), typeParameters.build()));
     }
 
-    /*
-     * Given signature with unmatched parameters return filtered one.
+    /**
+     * In case of coercion between un-parametrized types (DOUBLE, BIGINT, ...)
+     * we only need to know whether coercion is allowed (true|false).
+     * But in case of coercion to the type with long parameters, parameters of the result type must be calculated.
+     * Such coercion rules must be implemented within this method.
+     *
+     * Examples:
+     *
+     * LONG, DECIMAL(x,y) -> DECIMAL(19,0)
```

Comment:
It is needed for function that have parameters declared as a parametric types. 

If we try to add `BIGINT` and, say, `DECIMAL(2,1)` we need to find applicable function somehow. So the `resolveFunction` algorithm must pick the `ADD(DECIMAL(p1,s1), DECIMAL(p2,s2))` function. In order to do that we need to have `canCoerce()` that return `true` for the`BIGINT`,`DECIMAL(p1,s1)` types pair.  Besides picking the applicable function the actual types after coercion must be calculated. So the planner can insert appropriate casts. That's why we need `getCommonSupperType` to handle parametric signatures.

I know that it looks strange. I thought about creating separate methods that handle parametric signatures, called, say `canBindTypeToSingature` or `bindTypeToSignature`. But it seems that it is going to have almost the same algorithm. But now i'm thinking that maybe it make sense. What do you think?